### PR TITLE
Images can be attached to auctions

### PR DESCRIPTION
**Pull request by seoulection**
The functionality to attach images to auctions has not been fully tested. We need assistance for these tests. Does anyone have experience testing image attachments?

### CODE REVIEW

File: ebai-front/tests/unit/NumberInput.spec.js [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+        required: false,
+        id: "amount"
+      }
+    })
+    const label = wrapper.find('#amount')
+
+    expect(label.text()).toBe('Amount:')
+  })
+
+  it('can be required', () => {
```
Comment (omarnyte) [inline ebai-front/tests/unit/NumberInput.spec.js]: Can you also add a spec for the expected behavior when `required` is false? I find it helpful to test both conditions so that your tests can document expected behavior. I also find that those sorts of tests provide good feedback for helping you decide whether your component has become too complex. If you find that your tests have become deeply nested and complicated because of a lot of conditional props, it could be a sign that you should create a new component rather than adding a new prop (which I'm not necessarily suggesting for this case--just a thought).

File: ebai-front/tests/unit/CreateAuctionForm.spec.js [hunk 2]
```diff
@@ -50,4 +50,10 @@ describe('CreateAuctionForm.vue', () => {
     expect(button.exists()).toBe(true)
     expect(button.text()).toEqual('Create')
   })
+
+  it('has a file uploader', () => {
+    const fileInput = wrapper.find('#auctionImage')
+
+    expect(fileInput.exists()).toBe(true)
+  })
```
Comment (omarnyte) [inline ebai-front/tests/unit/CreateAuctionForm.spec.js]: As we discussed earlier, I haven't encountered a situation in which I've needed to test a feature to attach images. I think starting with a unit test is a good idea for now, but I'd actually go beyond just checking for the presence of *some* element with an id of `auctionImage"`. Particularly because of the lack of an integration test, I'd specifically look for an `input` element with a `type` of `file`. Even if this spec doesn't properly test that your form gets `POST`ed correctly, it at least ensures that the user can upload the image. 

It appears that [uploading a file fires off an `event`](https://developer.mozilla.org/en-US/docs/Web/API/HTMLElement/change_event), not unlike most other interactions with an `input`. I haven't taken a look at that specific event, but I'd begin by logging that event and seeing whether it's something that you could feasibly spy on or stub.

File: ebai-back/db/migrate/20200204171433_create_active_storage_tables.active_storage.rb [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+      t.string   :checksum,   null: false
+      t.datetime :created_at, null: false
+
+      t.index [ :key ], unique: true
+    end
+
+    create_table :active_storage_attachments do |t|
+      t.string     :name,     null: false
+      t.references :record,   null: false, polymorphic: true, index: false
+      t.references :blob,     null: false
```
Comment (kkrull) [inline ebai-back/db/migrate/20200204171433_create_active_storage_tables.active_storage.rb]: Storing the image as a blob in the database can work; it's an approach that is often used.  Keep in mind also the possible alternate design: store the image as a file on the file system and add a reference to that file (the path to that file) to the database.

There's an interesting discussion here: https://stackoverflow.com/questions/3748/storing-images-in-db-yea-or-nay

Starting with the images as blobs in the database is probably a good way to go, when there aren't many users or many/large images.  Deleting an auction also deletes the image so everything stays in sync.  Putting images on the filesystem introduces the need for backing up the files and the database, and adding application code to handle unexpected cases where the auction exists in the database, but the image file can't be found.

TL;DR - this looks fine; just want you to be aware of some of the factors that go into the decision.

File: ebai-front/tests/unit/CreateAuctionForm.spec.js (hunk 2 above)
Comment (kkrull) [reply ebai-front/tests/unit/CreateAuctionForm.spec.js]: This is another good example of the "Test granularity problem" we're discussing in the earlier PR -- should this behavior be tested as a single integration test, or 2 unit tests:

* one that the front-end makes a multi-part POST request and 
* a second one that the back end properly responds to and creates the appropriate side effects when receiving that kind of request

Note the subtle difference between the claims you can make, based upon the strategy:

* integration test: it works because I just showed you (empirical proof)
* disjoint unit tests: it works because the front-end makes a request that the back-end is prepared to receive _and_ the back-end sends a response that the front-end can process correctly.  This strategy can be lead faster-executing and perhaps simpler tests, but the assumptions (the front-end works if the back-end works like I thought; the back-end works if the front-end works like I thought) can break down over time and introduce API Drift.

TL;DR - I'd writing the simplest test(s) you can around this using whichever style seems doable.  Remember you don't have to dig into asserting the exact, binary content of the image to test that you're sending an image.  Ensuring that you send the right content-type and some non-zero number of bytes in the request body should be enough to prove beyond a reasonable doubt that you're making the intended request.
